### DISCUSSION TITLE
Need tips for unit test creation

### DISCUSSION THREAD

**Discussion by c0nwild** [elheck/Squiddrone#299] [Coding] [answered]
In my driver implementation there are a lot of functions following this template:

```
auto NRF24L01::SetPowerState(State power_state) noexcept -> types::DriverStatus {
  utilities::Byte config_reg(ReadRegister(reg::config::REG_ADDR));
  if (power_state == State::enabled) {
    config_reg.SetBit(reg::config::PWR_UP);
  }
  if (power_state == State::disabled) {
    config_reg.ClearBit(reg::config::PWR_UP);
  }

  return (WriteRegister(reg::config::REG_ADDR, config_reg.Get()));
}
```

Now what would you consider worth testing? I could mock the WriteRegister function and check if the correct bit was set but that´s about all I came up with...

**Comment by elheck** [comment]
What permutation of power_state triggers which return? You could make equivalency classification for this.

If WriteRegister is a private method of the same class, it should not be mocked.

**Reply by c0nwild** [reply]
Thing is, I thought returning the result of the SPI transfer would be nice to capture transmission errors. So the return value of WriteRegister is "success" or error of some kind.

**Reply by elheck** [reply]
This sounds to me more like a integration test case. Am I right with this assumption? You want to integrate SPI transfer into the test of the NRF...

**Comment by ThorsHamster** [comment, marked as answer]
If you say, this is a template you use several times inside of your module. Would it make sense, to create an internal utility function with this kind of logic and test this with several Usecases? Then you have only to use it here.

Otherwise I would say, you should test for the intended behaviour. So If the User of the Interface expects a SetPowerState this should also be tested. Therefore you would have to test, if the correct Bit inside of the Register is set. But this could lead to flaky Tests.

**Reply by c0nwild** [reply]
This might be a good idea. ATM i am hiding a complete protocol layer inside my driver function. I will try and implement a protocol stack for the master-slave communication.
